### PR TITLE
Remove DEGREE_MAX; rely on deal.II's facilities for pre-compiled code

### PR DESCRIPTION
**Pull request by kronbichler**
As discussed in https://github.com/exadg/exadg/issues/241#issuecomment-1176238309, this PR proposes to control the maximal compiled polynomial degree at the configure time of deal.II. This helps avoiding #241, but we also need support functionality as we seemed to have converged towards raising errors in case the user requests a degree we do not compile for.

One thing we might need to decide upon is whether we really want to delete `config.h` - without the degree variable, there is nothing left in there, but maybe we will add it again? Note that the applications I had to touch were due to this change. Right now, this is up to discussion: Do we really want to go down this route, and what more needs to be done?

The main difficulty left (that I do not have time to work on for the next week) is how to detect what degrees deal.II actually provides to us. If we want to disable tests that do fall outside the given range of supported degrees, we would need to run a code at configure time (of ExaDG) to see for which degree `FEEvaluation::fast_evaluation_supported(degree, degree+1)` returns false for the first time, as deal.II only gives us that information at run time, and propagate that into the ExaDG configuration, namely to the place where we decide which tests to run. All assertions inside ExaDG would be fine, as we can assert the run-time value of `CellIntegrator::fast_evaluation_supported(degree)`.

There is an additional catch with this, namely the fact that this `FE_EVAL_FACTORY_DEGREE_MAX` is an internal macro in deal.II that is not guaranteed to be stable or doing what we want in the long run. We would have had that problem already before, but the current solution feels like a hack because we add this define flag to all compilation units in deal.II. Maybe at some time we want to make this an actual configure inside deal.II, like we had in ExaDG? It's not something that helps us entirely here, because we want to support dealii-9.4 for a while, so we need a solution for that as well. (And there is no possibility to add a configure script to a 9.4.x release, as those are only for bug-fixes.)

### CODE REVIEW

File: include/exadg/utilities/resolution_parameters.h [hunk 1]
```diff
@@ -51,12 +48,12 @@ struct SpatialResolutionParameters
       prm.add_parameter("DegreeMin",
                         degree_min,
                         "Minimal polynomial degree of shape functions.",
-                        dealii::Patterns::Integer(1,EXADG_DEGREE_MAX),
+                        dealii::Patterns::Integer(),
```
[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/utilities/resolution_parameters.h:51]: Yes, 1 as minimum sound good. We can not expect that our discretizations in ExaDG are consistent for degree 0 (e.g. SIPG method), so I do not see a restriction by choosing 1 as the minimal degree.

**Comment by nfehn** [comment]
Many thanks @kronbichler . I think all we need for merging is that you can confirm that the tests are passing on your machine.

In my opinion, we should assert that pre-compiled versions are used. But I think this could be done also later. The important point to make sure is that the code does not longer crash, and I think this should be fulfilled by the present PR?

I also agree that it would make sense to have this parameter as a configure in deal.II.

**Comment by kronbichler** [comment]
> In my opinion, we should assert that pre-compiled versions are used. But I think this could be done also later. The important point to make sure is that the code does not longer crash, and I think this should be fulfilled by the present PR?

I think that is a good default setting, as we discussed in https://github.com/exadg/exadg/issues/241#issuecomment-1176222423. From a developer point of view, I would like to be able to quickly test all of ExaDG without compiling all degrees from deal.II, but that can then be the follow-up to the follow-up :wink:

**Comment by nfehn** [comment]
@kronbichler What about the do-not-merge label? Please indicate once this PR is ready to merge from your side.

**Comment by kronbichler** [comment]
I forgot that. It is ready from my end, possible extensions can be made later.

**Review by peterrum — APPROVED** [review]

**Pull request merged**